### PR TITLE
GraphQL API

### PR DESCRIPTION
**Pull request by a-rmz**
## Description :sparkles:
* Add a first version of a GraphQL API, to be used in Vene to query existing notifications
* Add a tests for the API
(The full description can be found on the related Jira ticket)

### Closes :no_good_woman:
**[VEN-674](https://helsinkisolutionoffice.atlassian.net/browse/VEN-674):** Existing notification templates could be queried via GQL API

### Automated tests :gear:️
```shell
$ pytest tests/test_notification_queries.py
```

### Manual testing :construction_worker_man:
1. Add `django-ilmoitin.api.schema.Query` to your `Schema` queries
2. From a GQL client, execute the following query:
```graphql
query Notifications {
  notificationTemplates {
    edges {
      node {
        id
        preview
        translations {
          subject
          languageCode
          bodyHtml
          bodyText
        }
      }
    }
  }
}
```

### CODE REVIEW

**Comment by a-rmz** [comment]
Not sure if there should be a version bump  🤔

File: django_ilmoitin/api/schema/types.py
Comment (igordavydsson) [inline django_ilmoitin/api/schema/types.py:20]: it doesn't need a resolver on this type? 🤔

Comment (a-rmz) [reply django_ilmoitin/api/schema/types.py:20]: WHOOPS that actually shouldn't be there. 😅 

Or should it... 🤔 Would it make sense to have a preview for every possible language instead of a "global" notification-level preview?

Comment (igordavydsson) [reply django_ilmoitin/api/schema/types.py:20]: I think it would be beneficial to have a preview for every language. and then global-level preview would be for the default language 🤔

**Comment by igordavydsson** [comment]
yeah, version bump would be called for after this 👍

File: tests/conftest.py
Comment (igordavydsson) [inline tests/conftest.py]: what? "fr"? :pepehmm:

Comment (a-rmz) [reply tests/conftest.py]: Ahhh le good ole' copy-paste.

File: django_ilmoitin/api/schema/types.py
Comment (igordavydsson) [inline django_ilmoitin/api/schema/types.py]: why do you set this to `True`? 🤔

Comment (a-rmz) [reply django_ilmoitin/api/schema/types.py]: Apparently it's not needed anymore 🤔 It was causing some error when trying to query saying that there was no connection, but apparently now that `NotificationTemplate` implements the `relay.Node` interface, that's solved.

Comment (igordavydsson) [reply django_ilmoitin/api/schema/types.py]: ah, alright 👍

**Review by igordavydsson — COMMENTED** [review]
Looks good to me! Good job! 👏 

Would be nice to get some fresh eyes on this 😅  Maybe @pjlampil @charn @quyenlq have time?

File: setup.py
Comment (charn) [inline setup.py]: I'm wondering if we want to force graphene/graphql as a requirement or would using the `extras_require` feature be a good way to go here? Not sure what other changes that might require for the implementation. :shrug: [[Declaring “Extras” (optional features with their own dependencies](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies)] Also not sure if this optional requirements thing is a good way to go, but just thinking out loud.

File: django_ilmoitin/apps.py
Comment (charn) [inline django_ilmoitin/apps.py]: Just out of curiosity, why is this import required here? I think the graphql types in this app would get automatically registered by just subclassing `django_ilmoitin.api.schema.Query` in the app's Query where `django_ilmoitin` package is used.

File: setup.py
Comment (a-rmz) [reply setup.py]: The `extras` sound like the reasonable approach, considering that
> Sometimes a project has “recommended” dependencies, that are not required for all uses of the project

which seems to be the exact case. I think as long as we document clearly on the README that there's the API and what it requires to use it (i.e. installing the extra deps), we should be good.

File: django_ilmoitin/apps.py
Comment (a-rmz) [reply django_ilmoitin/apps.py]: In my mind it made sense to me as well, but without manually "loading" the schema package beforehand, I kept bumping into this error:
image

Comment (igordavydsson) [reply django_ilmoitin/apps.py]: ok, this import during `ready` is not actually required, if in the project you do `import django_ilmoitin.api` or something similar. resolving this thread.

File: setup.py
Comment (igordavydsson) [reply setup.py]: Alex has now implemented graphql support as `extras` 👍

File: README.md
Comment (igordavydsson) [inline README.md]: should we enumerate the steps if there is only 1? 🙂

Comment (igordavydsson) [reply README.md]: I think the first step is installing the extras, and then this is the second step. maybe put them together in README 🙂

File: README.md
Comment (tuomas777) [inline README.md]: Typo `GrapQL`

File: README.md
Comment (tuomas777) [inline README.md:131]: If one wants the API enabled but is not familiar with Graphene, this will probably look pretty cryptic. Also, she cannot just copy-paste the example to get a working API, but will need all kinds of other Graphene specific stuff to be set up too. So at minimum, I would add a remark about Ilmoitin using Graphene for the API, and a link to Graphene setup docs.

Ideally, I think it should be possible to also enable the API easily without knowing or caring about Graphene at all, with just something like 

``` python
# project/urls.py
...
urlpatterns += path('ilmoitin/', include('ilmoitin.urls'))
```
Dunno if implementing that is actually worth the trouble though, probably not at least right now 🙂

File: tests/conftest.py
Comment (tuomas777) [inline tests/conftest.py]: Could use `notification_type` also here.

File: tests/conftest.py
Comment (tuomas777) [inline tests/conftest.py:27]: Not related to this PR in any way, but I wonder if this is actually used for something, or could it be just removed 🙂

File: tests/test_notification_queries.py
Comment (tuomas777) [inline tests/test_notification_queries.py:23]: Translation specific previews should probably be tested as well (?)

**Review by tuomas777 — COMMENTED** [review]
Added a couple of comments. Good stuff! 👍

File: README.md
Comment (igordavydsson) [reply README.md]: :derp:

Comment (a-rmz) [reply README.md]: :derp: x2

Comment (igordavydsson) [reply README.md]: grapql
image

Comment (pjlampil) [reply README.md]: 🍿

File: tests/conftest.py
Comment (quyenlq) [inline tests/conftest.py:25]: Not sure the purpose of assigning "event_created" to a variable. I couldn't see any other use of `notification_type` or am I missing anything here

File: tests/conftest.py
Comment (a-rmz) [reply tests/conftest.py]: 😛 good catch

File: tests/conftest.py
Comment (a-rmz) [reply tests/conftest.py:27]: Hmm yeah, I think it's not being used for anything, so it's safe to delete. 👍

File: tests/conftest.py
Comment (a-rmz) [reply tests/conftest.py:25]: Just so that the places where `"event_created"` is required, a single value coming from the variable is used instead of having it hard-coded in different places.

File: tests/test_notification_queries.py
Comment (a-rmz) [reply tests/test_notification_queries.py:23]: Good catch!

File: README.md
Comment (a-rmz) [reply README.md]: image

File: README.md
Comment (a-rmz) [reply README.md:131]: Fair point. I think the easiest option would be to add some more details on the docs about the API requiring a working `graphene` API and link to their docs.

The logic we were following for this was to have a separate "plug-n-play" schema which you could mount in an existing GraphQL (graphene) API, so you wouldn't need to have a different endpoint _just_ for `ilmoitin`.

This implementation doesn't have everything it requires to have a standalone Graphene server, so it would require more work to have the `path('ilmoitin/', include('ilmoitin.urls'))` which would have the API out of the box.

cc @igordavydsson

Comment (igordavydsson) [reply README.md:131]: yeah, very good idea to have the fully functional API within ilmoitin, that you could just connect to a url path within your project 👍  it would indeed require a bit more work though, so maybe we could put it on hold for now. I think the docs that Alex added point the potential user in the right direction of installing graphene and setting up a working API first, so that's probably enough for now. 🙂

File: tests/conftest.py
Comment (igordavydsson) [reply tests/conftest.py:27]: yes, probably can be deleted

File: tests/conftest.py
Comment (igordavydsson) [reply tests/conftest.py:25]: now that it used a few lines below, makes at least some sense to have it in a var 😅  when the var was not used at all, it was a bit silly indeed 😬

File: tests/conftest.py
Comment (igordavydsson) [reply tests/conftest.py:27]: I will delete it after I merge the PR 👍

**Review by igordavydsson — APPROVED** [review]
Good job @a-rmz 👍 🎖️ 🥇 
Thanks guys for the comments, @charn @quyenlq @tuomas777 @pjlampil very much appreciated 🤗 

good job guys

**Pull request merged**
